Network/Socket/Buffer.hs

```diff
@@ -85,29 +89,22 @@ sendBuf sock str len =
 --
 -- NOTE: blocking on Windows unless you compile with -threaded (see
 -- GHC ticket #1129)
-recvBufFrom :: Socket -> Ptr a -> Int -> IO (Int, SockAddr)
-recvBufFrom sock@Socket{..} ptr nbytes
+recvBufFrom :: SocketAddress sa => Socket -> Ptr a -> Int -> IO (Int, sa)
+recvBufFrom s ptr nbytes
  | nbytes <= 0 = ioError (mkInvalidRecvArgError "Network.Socket.recvBufFrom")
- | otherwise   =
-    withNewSockAddr socketFamily $ \ptr_addr sz ->
+ | otherwise   = do
+    let fd = fdSocket s
+    withNewSocketAddress $ \ptr_sa sz ->
       alloca $ \ptr_len -> do
         poke ptr_len (fromIntegral sz)
-        len <- throwSocketErrorWaitRead sock "Network.Socket.recvBufFrom" $
-                   c_recvfrom socketFd ptr (fromIntegral nbytes) 0{-flags-}
-                                ptr_addr ptr_len
+        len <- throwSocketErrorWaitRead fd "Network.Socket.recvBufFrom" $
+                   c_recvfrom fd ptr (fromIntegral nbytes) 0{-flags-}
+                                ptr_sa ptr_len
         let len' = fromIntegral len
         if len' == 0
          then ioError (mkEOFError "Network.Socket.recvFrom")
          else do
-           flg <- isConnected sock
-             -- For at least one implementation (WinSock 2), recvfrom() ignores
-             -- filling in the sockaddr for connected TCP sockets. Cope with
-             -- this by using getPeerName instead.
-           sockaddr <-
-                if flg then
-                   getPeerName sock
-                else
-                   peekSockAddr ptr_addr
+           sockaddr <- getPeerName s -- xxx fixme: peekSockAddr ptr_sa
```

Comment:
This cannot be implemented without `SocketStatus`. 

```
             -- For at least one implementation (WinSock 2), recvfrom() ignores
             -- filling in the sockaddr for connected TCP sockets. Cope with
             -- this by using getPeerName instead.
```

@Mistuke Is this comment correct even at this moment?

Reply:
https://msdn.microsoft.com/ja-jp/library/windows/desktop/ms740120(v=vs.85).aspx

> The from and fromlen parameters are ignored for connection-oriented sockets.

Wow! But it is very rare to use `recvfrom()` for TCP. So, is documentation good enough?

Reply:
Manual on macOS:

> If address is not a null pointer and the socket is not connection-oriented, the source address of the message is filled in.  The address_len argument is a value-result argument, initialized to the size of the buffer associated with address, and modified on return to indicate the actual size of the address stored there.

Reply:
Done.

Reply:
@kazu-yamamoto yeah, because Winsock is designed to be Berkeley sockets complaint but not POSIX. I believe Berkeley also doesn't support `recvfrom` on connected sockets http://wiki.treck.com/Introduction_to_BSD_Sockets#recvfrom.28.29

Reply:
@Mistuke Thank you!